ragna/source_storages/_lancedb.py

```diff
@@ -81,6 +94,7 @@ def store(
                             "id": str(uuid.uuid4()),
                             "document_id": str(document.id),
                             "document_name": str(document.name),
+                            **document.metadata,
```

Comment:
Currently lancedb doesn't allow partial table updates, so if we create a schema with all the possible metadata fields, each insertion needs to contain every field.

@pmeier Do you have a preference between these options?
1. Create the table schema at runtime with `id`, `document_id`, `document_name`, `page_numbers`, `text`, and `metadata`, where `metadata` is some kind of struct or serialized dict. Then `_translate_metadata_filter` will need to account for some of the fields being nested in the `metadata` column. Pros: can create the schema at runtime, can call `.store()` multiple times without rebuilding the schema. Cons: some potentially messy nesting.
2. Create top-level fields from the union of keys from the document metadatas and build the schema within `.store()` (this is what I have here). Cons: need to build out empty values for documents that come in with missing metadata fields because LanceDB won't accept missing fields.
3. Just create a simple schema as in Option 1 but exclude the `metadata` field. Create the table from this basic schema, then when a document comes in with additional metadata, add new columns for the metadata. Pros: matches most closely the behavior of Chroma. Cons: possibly slow unless [`LanceTable.add_columns`](https://github.com/lancedb/lancedb/blob/513926960db94f534f3fee69fafbc02d0e576488/python/python/lancedb/table.py#L1740) is okay to run a lot.

cc @nenb

Reply:
IMO

1. Unless there is some magic that Lance readily offers, this will be a nightmare to implement. Imagine metadata being a serialized JSON object. How are we going to effectively filter on the individual fields?
2. / 3. I'm unsure what the difference is. We should probably treat the first document that we see as "prototype" and build the table with its metadata. In case all documents have the same metadata, which is a reasonable assumption for >= 80% of uses cases, this requires no extra work. In case the assumption doesn't hold, we'll have to
    1. fill missing metadata fields with `None` or whatever is appropriate
    2. `add_columns` in case we encounter new metadata.

Does that make sense?

Reply:
Definitely makes sense. I'll go ahead and make this change. My remaining question is do we ever expect users to call `.store()` multiple times and, if so, should the method
1. append
2. overwrite
3. offer a `overwrite: bool = False` param?

Reply:
As discussed, we assume 1. for the builtin source storages.